src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8389,40 +8469,179 @@ protected MethodGroupResolution BindExtensionMethod(
+                    options, returnRefKind, returnType, in callingConvention, diagnostics, scope,
+                    ref actualArguments, out MethodGroupResolution extensionMethodResult))
+                {
+                    // If the search in the current scope resulted in any applicable method (regardless of whether a best
+                    // applicable method could be determined) then our search is complete. Otherwise, store aside the
+                    // first non-applicable result and continue searching for an applicable result.
+                    if (extensionMethodResult.HasAnyApplicableMethod)
+                    {
+                        firstResult.Free();
+                        actualArguments?.Free();
+                        diagnostics.Free();
+                        return extensionMethodResult;
+                    }
+                    else if (firstResult.IsEmpty)
+                    {
+                        firstResult = extensionMethodResult;
+                    }
+                    else
+                    {
+                        // Neither the first result, nor applicable. No need to save result.
+                        extensionMethodResult.Free();
+                    }
+                }
+            }
+
+            Debug.Assert((actualArguments == null) || !firstResult.IsEmpty);
+            actualArguments?.Free();
+            diagnostics.Free();
+
+            return firstResult;
+
+            static bool tryResolveExtensionMember(Binder binder, SyntaxNode expression, string memberName, AnalyzedArguments? analyzedArguments, BoundExpression left,
+                ImmutableArray<TypeWithAnnotations> typeArgumentsWithAnnotations, OverloadResolution.Options options, RefKind returnRefKind, in CallingConventionInfo callingConvention,
+                TypeSymbol? returnType, BindingDiagnosticBag diagnostics, ExtensionScope scope, out MethodGroupResolution extensionResult)
+            {
+                var lookupResult = LookupResult.GetInstance();
+
+                int arity = typeArgumentsWithAnnotations.IsDefault ? 0 : typeArgumentsWithAnnotations.Length;
+                var lookupOptions = (arity == 0) ? LookupOptions.AllMethodsOnArityZero : LookupOptions.Default;
+                if (analyzedArguments is not null)
+                {
+                    lookupOptions |= LookupOptions.MustBeInvocableIfMember;
+                }
 
-            foreach (var scope in new ExtensionMethodScopes(this))
+                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = binder.GetNewCompoundUseSiteInfo(diagnostics);
+
+                scope.Binder.LookupExtensionMembersInSingleBinder(
+                    lookupResult, left.Type!, memberName, arity,
+                    basesBeingResolved: null, lookupOptions, originalBinder: binder, ref useSiteInfo);
+
+                diagnostics.Add(expression, useSiteInfo);
```

Comment:
>diagnostics.Add(expression, useSiteInfo); [](http://example.com/codeflow?start=16&length=41)

I think this is where the problem is. This is where we are dropping use-site in case diagnostics is dropped below. And the diagnostics is dropped for a good reason. I think use-site diagnostics should be collected separately on the ```ResolveExtension``` level. I.e. it should be passed in as a parameter and we shouldn't be merging it into diagnostic bags that might be dropped.